pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -121,6 +121,32 @@ public List<String> getPrimaryKeyColumns() {
     return _primaryKeyColumns;
   }
 
+  private MutableRoaringBitmap getQueryableDocIds(IndexSegment segment, String deleteRecordColumn,
+      MutableRoaringBitmap validDocIds) {
+    MutableRoaringBitmap queryableDocIds = null;
+    if (deleteRecordColumn != null) {
+      queryableDocIds = new MutableRoaringBitmap();
+      // we allow passing null values for primaryKey columns and comparisonColumns to read deleteColumn to avoid paying
+      // the overhead of reading primary key, comparison column etc. They will never be null in other scenarios.
+      UpsertUtils.RecordInfoReader recordInfoReader =
```

Comment:
Suggest not modifying the `RecordInfoReader` because it is not designed for this purpose. We can simply create a `PinotSegmentColumnReader` for `_deleteRecordColumn`, then iterate over the `validDocIds`

Reply:
when iterate the value of _deleteRecordColumn, can deletedRecord be null? can this be further optimize?
```
        if (deleteRecordColumnReader.isNull(doc_id)) {
          // defaultNullValue for deleteColumn is "false"
          queryableDocIds.add(doc_id);
        } else if (!(Boolean) deleteRecordColumnReader.getValue(doc_id)) {
          queryableDocIds.add(doc_id);
        }
```